### PR TITLE
MF-1197 - Add MQTT adapter tests

### LINKED ISSUE #1197: Add MQTT adapter tests
**FEATURE REQUEST**

2. Describe the feature you are requesting, as well as the possible use case(s) for it.
MQTT adapter needs unit tests.

3. Indicate the importance of this feature to you (must-have, should-have, nice-to-have).
This is a must-have.

### PR DESCRIPTION
**Pull request by zzokki81**
### What does this do?
Added MQTT unit tests

### Which issue(s) does this PR fix/relate to?
Resolves #1197 

### Have you included tests for your changes?
MQTT unit tests

### Did you document any new/modified functionality?
No

### CODE REVIEW

File: mqtt/handler_test.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+func TestHandler_AuthConnect(t *testing.T) {
+
+	cases := []struct {
+		desc    string
+		err     error
+		session *session.Client
+	}{
+		{
+			desc:    "should return error if client is nil",
```
Comment (manuio) [inline mqtt/handler_test.go]: The description should explain the test that you are doing not the error that is supposed to return. In this case it can be `Handle message without active session`.

Same for all other description

File: mqtt/mocks/client.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+func NewClient() auth.Client {
+	return MockClient{}
+}
+
+func (cli MockClient) Authorize(ctx context.Context, chanID, thingID string) error {
+	return nil
+}
+
+func (cli MockClient) Identify(ctx context.Context, thingKey string) (string, error) {
+
```
Comment (manuio) [inline mqtt/mocks/client.go]: kill this blank line

**Review by manuio — CHANGES_REQUESTED** [review]

File: mqtt/handler_test.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	rdb "github.com/go-redis/redis/v8"
+	"github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/mqtt/mocks"
+	"github.com/mainflux/mainflux/mqtt/redis"
+	"github.com/mainflux/mainflux/pkg/messaging"
+	"github.com/ory/dockertest/v3"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestHandler_AuthConnect(t *testing.T) {
```
Comment (manuio) [inline mqtt/handler_test.go]: Rename the method `TestAuthConnect`

File: mqtt/handler_test.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+		},
+	}
+
+	for _, tc := range cases {
+		err := setupTest().AuthConnect(tc.session)
+		assert.True(t, errors.Contains(err, tc.err), fmt.Sprintf("%s: expected %s got %s\n", tc.desc, tc.err, err))
+	}
+}
+
+func TestHandler_AuthPublish(t *testing.T) {
```
Comment (manuio) [inline mqtt/handler_test.go]: Rename `TestAuthPublish`

**Review by manuio — CHANGES_REQUESTED** [review]

File: mqtt/handler_test.go [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+		{
+			desc:   "Handle subscribe authorization on client correctly",
+			client: &sessionClient,
+			err:    nil,
+			topic:  &topics,
+		},
+	}
+
+	for _, tc := range cases {
+		err := setupTest().AuthSubscribe(tc.client, tc.topic)
```
Comment (manuio) [inline mqtt/handler_test.go]: Create the session handler outside the for loop: `handler := createHandler()`

File: mqtt/handler_test.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+	}
+
+	cases := []struct {
+		desc   string
+		client *session.Client
+		err    error
+		topic  *[]string
+	}{
+		{
+			desc:   "Handle subscribe authorization on client without active session",
```
Comment (manuio) [inline mqtt/handler_test.go]: Avoid repeting `Handle subscribe authorization on client` for all descriptions. For example here `Subscribe without active session` would be enough

**Review by manuio — CHANGES_REQUESTED** [review]

File: mqtt/handler_test.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+import (
+	"fmt"
+	"github.com/mainflux/mainflux/pkg/uuid"
+	"github.com/stretchr/testify/require"
+	"log"
+	"testing"
+
+	"github.com/mainflux/mainflux/pkg/errors"
+
```
Comment (manuio) [inline mqtt/handler_test.go]: remove this blank line

File: mqtt/handler_test.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+			desc:    "connect without active session",
+			err:     errInvalidConnect,
+			session: nil,
+		},
+		{
+			desc: "connect when id is invalid",
+			err:  errors.New("thing identify error"),
+			session: &session.Client{
+				ID:       "123",
+				Username: "testUsername",
```
Comment (manuio) [inline mqtt/handler_test.go]: create a global const and use it in all tests

**Review by manuio — CHANGES_REQUESTED** [review]

File: mqtt/handler_test.go [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+		err     error
+		session *session.Client
+	}{
+		{
+			desc:    "connect without active session",
+			err:     errInvalidConnect,
+			session: nil,
+		},
+		{
+			desc: "connect when id is invalid",
```
Comment (manuio) [inline mqtt/handler_test.go]: `connect with invalid id`

File: mqtt/handler_test.go [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+			err:  errors.New("thing identify error"),
```
Comment (manuio) [inline mqtt/handler_test.go]: Remove this error and use one from the package. If there is no error in the package maybe you can change a bit the code and wrap the returned error with a new one.

File: mqtt/handler_test.go [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+			session: &session.Client{
+				ID:       "123",
+				Username: "testUsername",
+				Password: []byte(""),
+			},
+		},
+		{
+			desc: "connect when username is invalid",
```
Comment (manuio) [inline mqtt/handler_test.go]: `connect with valid password and invalid username`

File: mqtt/handler_test.go [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+			err:  errors.ErrAuthentication,
+			session: &session.Client{
+				ID:       "123",
+				Username: "testUsername",
+				Password: []byte("password"),
+			},
+		},
+		{
+			desc: "connect with right username and password",
```
Comment (manuio) [inline mqtt/handler_test.go]: `connect with valid username and password"`

**Review by manuio — CHANGES_REQUESTED** [review]

File: mqtt/handler_test.go [hunk 13]
```diff
@@ -2,21 +2,26 @@ package mqtt
...
 	"github.com/mainflux/mainflux/mqtt/redis"
+	"github.com/mainflux/mainflux/pkg/errors"
 	"github.com/mainflux/mainflux/pkg/messaging"
+	"github.com/mainflux/mainflux/pkg/uuid"
 	"github.com/mainflux/mproxy/pkg/session"
 	"github.com/ory/dockertest/v3"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"log"
+	"testing"
```
Comment (manuio) [inline mqtt/handler_test.go]: Move this 2 packets to the top

File: mqtt/handler_test.go [hunk 14]
```diff
@@ -2,21 +2,26 @@ package mqtt
...
+)
+
+const (
+	id         = "123"
+	username   = "testUsername"
+	usernameOK = "ok"
```
Comment (manuio) [inline mqtt/handler_test.go]: Why not to use a single user for the tests?

**Review by manuio — CHANGES_REQUESTED** [review]

File: mqtt/handler_test.go [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+	for _, tc := range cases {
+		handler.Publish(tc.client, &tc.topic, &tc.payload)
+		assert.Contains(t, buf.String(), tc.expected)
+	}
+}
+
+func newHandler() session.Handler {
+	logger, err := logger.New(&buf, "debug")
+	if err != nil {
+		panic(err)
```
Comment (manuio) [inline mqtt/handler_test.go]: Replace it by something like `log.Fatalf("Could not create logger: %s", err))`

**Review by manuio — CHANGES_REQUESTED** [review]

File: mqtt/handler_test.go [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+	"github.com/mainflux/mainflux/pkg/uuid"
+	"github.com/mainflux/mproxy/pkg/session"
+	"github.com/ory/dockertest/v3"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	id              = "123"
+	username        = "username"
```
Comment (manuio) [inline mqtt/handler_test.go]: Since we are using thingID as MQTT username it would be better to use an uuid here:
```
	thingID  = "thingID"
```

File: mqtt/handler_test.go [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+	invalidUsername = "invalidUsername"
```
Comment (manuio) [inline mqtt/handler_test.go]: `invalidID = "invalidID"`

File: mqtt/handler_test.go [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+	for _, tc := range cases {
+		err := handler.AuthConnect(tc.session)
+		assert.True(t, errors.Contains(err, tc.err), fmt.Sprintf("%s: expected %s got %s\n", tc.desc, tc.err, err))
+	}
+}
+
+func TestAuthPublish(t *testing.T) {
+	handler := newHandler()
+	var topic string
+	var payload = []byte("payload")
```
Comment (manuio) [inline mqtt/handler_test.go]: Use a valid payload. For example `[]byte("[{"n":"test-name", "v": 1.2}]")`

File: mqtt/handler_test.go [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+		desc    string
+		client  *session.Client
+		err     error
+		topic   *string
+		payload []byte
+	}{
+		{
+			desc:    "publish without active session",
+			client:  nil,
+			err:     errNilClient,
```
Comment (manuio) [inline mqtt/handler_test.go]: It's not realted to this PR but can you change this error name please? `errInactiveSession` for example

File: mqtt/handler_test.go [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+		log.Fatalf("Could not start container: %s", err)
+	}
+
+	redisClient := rdb.NewClient(&rdb.Options{
+		Addr:     fmt.Sprintf("localhost:%s", container.GetPort("6379/tcp")),
+		Password: "",
+		DB:       0,
+	})
+
+	eventStore := redis.NewEventStore(redisClient, "")
```
Comment (manuio) [inline mqtt/handler_test.go]: No need to create a real even store. Use a mock instead. The only method to implement is the one reaturning the client ID: `ID()`

**Review by manuio — CHANGES_REQUESTED** [review]

File: mqtt/handler_test.go [hunk 21]
```diff
@@ -0,0 +1,373 @@
...
+		err     error
+		session *session.Client
+	}{
+		{
+			desc:    "connect without active session",
+			err:     errMissingClient,
+			session: nil,
+		},
+		{
+			desc: "connect with valid id",
```
Comment (manuio) [inline mqtt/handler_test.go]: `connect with invalid password"` 

Maybe you can add another test with empty client ID?

File: mqtt/handler_test.go [hunk 22]
```diff
@@ -0,0 +1,373 @@
...
+
+	cases := []struct {
+		desc    string
+		client  *session.Client
+		err     error
+		topic   *string
+		payload []byte
+	}{
+		{
+			desc:    "publish without active session",
```
Comment (manuio) [inline mqtt/handler_test.go]: `publish with inactive client`

File: mqtt/handler_test.go [hunk 23]
```diff
@@ -0,0 +1,373 @@
...
+		},
+		{
+			desc:    "publish without topic",
+			client:  &sessionClient,
+			err:     errMissingTopicPub,
+			topic:   nil,
+			payload: payload,
+		},
+		{
+			desc:    "publish with active session, valid topic and payload",
```
Comment (manuio) [inline mqtt/handler_test.go]: `publish with malformed topic`

**Review by manuio — CHANGES_REQUESTED** [review]

**Review by manuio — CHANGES_REQUESTED** [review]

File: mqtt/handler_test.go [hunk 24]
```diff
@@ -0,0 +1,381 @@
...
+		desc     string
+		client   *session.Client
+		topic    string
+		payload  []byte
+		expected string
+	}{
+		{
+			desc:     "publish without active session",
+			client:   nil,
+			topic:    "topic",
```
Comment (manuio) [inline mqtt/handler_test.go]: Use a valid topic here to be sure that the error comes from the inactive session

File: mqtt/handler_test.go [hunk 25]
```diff
@@ -0,0 +1,381 @@
...
+			payload:  []byte("payload"),
+			expected: errMissingClient.Error(),
+		},
+		{
+			desc:     "publish with invalid channel parts",
```
Comment (manuio) [inline mqtt/handler_test.go]: This should be `publish with invalid topic`

File: mqtt/handler_test.go [hunk 26]
```diff
@@ -0,0 +1,381 @@
...
+			client:   &sessionClient,
+			topic:    "topic",
+			payload:  []byte("payload"),
+			expected: "published with client_id " + clientID + " to the topic: " + "topic",
```
Comment (manuio) [inline mqtt/handler_test.go]: Create a variable for this error

File: mqtt/handler_test.go [hunk 27]
```diff
@@ -0,0 +1,381 @@
...
+		desc     string
+		client   *session.Client
+		topic    []string
+		expected string
+	}{
+		{
+			desc:     "subscribe without active session",
+			client:   nil,
+			topic:    topics,
+			expected: "failed to subscribe: " + errMissingClient.Error(),
```
Comment (manuio) [inline mqtt/handler_test.go]: Create an error for `failed to subscribe:`

**Review by manuio — CHANGES_REQUESTED** [review]

File: mqtt/handler.go [hunk 28]
```diff
@@ -69,7 +77,7 @@ func (h *handler) AuthConnect(c *session.Client) error {
 	}
 
 	if err := h.es.Connect(c.Username); err != nil {
-		h.logger.Warn("Failed to publish connect event: " + err.Error())
+		h.logger.Warn("failed to publish connect event: " + err.Error())
```
Comment (manuio) [inline mqtt/handler.go]: I'd change this `Warn` by `Error` and create a const for the message

**Review by manuio — CHANGES_REQUESTED** [review]

File: mqtt/handler.go [hunk 29]
```diff
@@ -153,38 +161,38 @@ func (h *handler) Publish(c *session.Client, topic *string, payload *[]byte) {
...
 	if c == nil {
-		h.logger.Error("Nil client disconnect")
+		h.logger.Error(errors.Wrap(errFailedDisconnectMsg, errClientNotInitialized).Error())
 		return
 	}
-	h.logger.Info("Disconnect - Client with ID: " + c.ID + " and username " + c.Username + " disconnected")
+	h.logger.Info(fmt.Sprintf(infoDisconnectMsg, c.ID, c.Username))
 	if err := h.es.Disconnect(c.Username); err != nil {
-		h.logger.Warn("Failed to publish disconnect event: " + err.Error())
+		h.logger.Warn("failed to publish disconnect event: " + err.Error())
```
Comment (manuio) [inline mqtt/handler.go]: Same here. `logger.Error` and create a const for the message. Also, test it

**Review by manuio — CHANGES_REQUESTED** [review]

File: mqtt/handler_test.go [hunk 30]
```diff
@@ -0,0 +1,392 @@
...
+			err:     errClientNotInitialized,
+			session: nil,
+		},
+		{
+			desc: "connect without clientID",
+			err:  nil,
+			session: &session.Client{
+				ID:       "",
+				Username: thingID,
+				Password: []byte("password"),
```
Comment (manuio) [inline mqtt/handler_test.go]: Use `password` const here

File: mqtt/handler_test.go [hunk 31]
```diff
@@ -0,0 +1,392 @@
...
+		session *session.Client
+	}{
+		{
+			desc:    "connect without active session",
+			err:     errClientNotInitialized,
+			session: nil,
+		},
+		{
+			desc: "connect without clientID",
+			err:  nil,
```
Comment (manuio) [inline mqtt/handler_test.go]: Is it normal to don't return error if client ID is empty?

**Review by manuio — CHANGES_REQUESTED** [review]

File: mqtt/handler_test.go [hunk 32]
```diff
@@ -0,0 +1,392 @@
...
+
+	for _, tc := range cases {
+		err := handler.AuthConnect(tc.session)
+		assert.True(t, errors.Contains(err, tc.err), fmt.Sprintf("%s: expected %s got %s\n", tc.desc, tc.err, err))
+	}
+}
+
+func TestAuthPublish(t *testing.T) {
+	handler := newHandler()
+	var invalidTopic = "invalidTopic"
```
Comment (manuio) [inline mqtt/handler_test.go]: This var is used in several tests, create a const instead.

**Review by manuio — CHANGES_REQUESTED** [review]

File: mqtt/handler_test.go [hunk 33]
```diff
@@ -0,0 +1,406 @@
...
+	var topic = "channels/" + chID + "/messages"
+
+	cases := []struct {
+		desc    string
+		client  *session.Client
+		err     error
+		topic   *string
+		payload []byte
+	}{
+		{
```
Comment (manuio) [inline mqtt/handler_test.go:109]: `publish successfully` test is missing

File: mqtt/handler_test.go [hunk 34]
```diff
@@ -0,0 +1,406 @@
...
+		client   *session.Client
+		expected string
+	}{
+		{
+			desc:     "connect without active session",
+			client:   nil,
+			expected: fmt.Sprint(errors.Wrap(mqtt.ErrFailedConnect, mqtt.ErrClientNotInitialized).Error()),
+		},
+		{
+			desc:     "connect with valid session",
```
Comment (manuio) [inline mqtt/handler_test.go]: `connect with active session` would be more consistent with previous message

File: mqtt/handler_test.go [hunk 35]
```diff
@@ -0,0 +1,406 @@
...
+	buf.Reset()
+	cases := []struct {
+		desc     string
+		client   *session.Client
+		expected string
+	}{
+		{
+			desc:     "connect without active session",
+			client:   nil,
+			expected: fmt.Sprint(errors.Wrap(mqtt.ErrFailedConnect, mqtt.ErrClientNotInitialized).Error()),
```
Comment (manuio) [inline mqtt/handler_test.go]: use `err: errors.Wrap(mqtt.ErrFailedConnect, mqtt.ErrClientNotInitialized)` here and extract the string latter

File: mqtt/handler_test.go [hunk 36]
```diff
@@ -0,0 +1,406 @@
...
+	}
+}
+
+func TestConnect(t *testing.T) {
+	handler := newHandler()
+	buf.Reset()
+	cases := []struct {
+		desc     string
+		client   *session.Client
+		expected string
```
Comment (manuio) [inline mqtt/handler_test.go]: `err`

File: mqtt/handler_test.go [hunk 37]
```diff
@@ -0,0 +1,406 @@
...
+
+	for _, tc := range cases {
+		handler.Connect(tc.client)
+		assert.Contains(t, buf.String(), tc.expected)
+	}
+}
+
+func TestPublish(t *testing.T) {
+	handler := newHandler()
+	buf.Reset()
```
Comment (manuio) [inline mqtt/handler_test.go]: I'd rename `buf` `logBuffer`

**Review by manuio — CHANGES_REQUESTED** [review]

**Review by manuio — DISMISSED** [review]
LGTM

File: mqtt/handler_test.go [hunk 38]
```diff
@@ -0,0 +1,421 @@
...
+	authClient := mocks.NewClient(map[string]string{password: thingID}, map[string]interface{}{chanID: thingID})
+
+	var redisClient *redis.Client
+
+	pool, err := dockertest.NewPool("")
+	if err != nil {
+		log.Fatalf("Could not connect to docker: %s", err)
+	}
+
+	container, err := pool.Run("redis", "5.0-alpine", nil)
```
Comment (dborovcanin) [inline mqtt/handler_test.go]: Spinning a new Docker container for each `newHandler` call is not a good idea. Why do we even use Redis and not mock for events?

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Testing using log messages is usually not a good idea. While it makes sense for some methods, we definitely should not spin Redis for each `newHandler` call, and we should use an error message to compare (see error wrapping), not hardcoded log constants.

**Review by manuio — DISMISSED** [review]
LGTM

File: mqtt/handler.go [hunk 39]
```diff
@@ -23,14 +24,28 @@ var _ session.Handler = (*handler)(nil)
...
+	LogInfoPublished                   = "published with client_id %s to the topic %s"
+	LogErrFailedConnect                = "failed to connect: "
+	LogErrFailedSubscribe              = "failed to subscribe: "
+	LogErrFailedUnsubscribe            = "failed to unsubscribe: "
+	LogErrFailedPublish                = "failed to publish: "
+	LogErrFailedDisconnect             = "failed to disconnect: "
+	LogErrFailedPublishDisconnectEvent = "failed to publish disconnect event: "
+	logErrFailedParseSubtopic          = "failed to parse subtopic: "
+	LogErrFailedPublishConnectEvent    = "failed to publish connect event: "
+	LogErrFailedPublishToMsgBroker     = "failed to publish to mainflux message broker: "
```
Comment (dborovcanin) [inline mqtt/handler.go:40]: Put log messages to a separate `const` block.

File: mqtt/handler.go [hunk 40]
```diff
@@ -23,14 +24,28 @@ var _ session.Handler = (*handler)(nil)
...
+	errMalformedSubtopic               = errors.New("malformed subtopic")
```
Comment (dborovcanin) [inline mqtt/handler.go]: Is this error covered by tests?

File: mqtt/handler_test.go [hunk 41]
```diff
@@ -0,0 +1,391 @@
...
+	invalidChanID  = "1"
+	clientID       = "clientID"
+	invalidThingID = "invalidThingID"
+	password       = "password"
+)
+
+var (
+	invalidTopic  = "invalidTopic"
+	idProvider    = uuid.NewMock()
+	logBuffer     = bytes.Buffer{}
```
Comment (dborovcanin) [inline mqtt/handler_test.go:37]: Add a comment, something like:
Test log messages for cases the handler does not provide a return value.

**Review by dborovcanin — CHANGES_REQUESTED** [review]
This obscure testing of logs pointed out potential flows in mProxy. Also, we should definitely decouple mutual dependency between Mainflux and mProxy (mProxy is using Mainflux errors, while Mainflux is using mProxy in MQTT adapter).

**Review by manuio — DISMISSED** [review]
LGTM

File: mqtt/handler.go [hunk 42]
```diff
@@ -20,17 +21,33 @@ import (
 
 var _ session.Handler = (*handler)(nil)
 
-const protocol = "mqtt"
+const (
+	protocol                           = "mqtt"
```
Comment (dborovcanin) [inline mqtt/handler.go]: Keep `protocol` as a separate constant, since this is a block for logging constants.

**Review by dborovcanin — CHANGES_REQUESTED** [review]
While checking the log message is a quirky way to test - for some methods that don't return errors but log them - that's the only way. I left only one cosmetical remark and we can merge.

**Review by manuio — APPROVED** [review]
LGTM

**Review by dborovcanin — APPROVED** [review]

**Review by drasko — APPROVED** [review]
LGTM

**Comment by drasko** [comment]
Merged! Thanks @zzokki81. This was a nice solution with log analysis, but we should improve in the future by adding return values to the handlers.

**Comment by dborovcanin** [comment]
The main improvement would be adding return errors where it's missing, and using pubsub mocks to check if pubsub methods are called for successful flow (rather than checking for info logs). Also, moving the logger from the handler implementation to the logging wrapper/middleware is a good idea.

**Comment by drasko** [comment]
Issue to track: https://github.com/mainflux/mproxy/issues/31
